pkg/server/image_pull.go

```diff
@@ -17,14 +17,202 @@ limitations under the License.
 // PullImage pulls an image with authentication config.
+// TODO add authentication
+// TODO harden api
 func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
-	return nil, errors.New("not implemented")
+	var (
+		err  error
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	image := r.GetImage().Image
+
+	if desc, size, err = c.pullImage(image); err != nil {
+		return nil, fmt.Errorf("failed to pull image %q: %v", image, err)
+	}
+	digest := desc.Digest.String()
+
+	// TODO pass a metadata struct to pullimage
+	// store the image metadata
+	var meta *metadata.ImageMetadata
+	meta, err = c.imageMetadataStore.Get(digest)
+	if err == nil {
+		if meta == nil {
+			meta = &metadata.ImageMetadata{
+				ID:          digest,
+				RepoTags:    []string{image},       // TODO fill this array properly
+				RepoDigests: []string{digest},      // TODO fill this array properly
+				Size:        uint64(size),          // TODO  compressed or uncompressed size? using compressed
+				Auth:        r.GetAuth(),           // TODO security
+				CreatedAt:   time.Now().UnixNano(), // TODO or should this be the image created at time?
+			}
+			if err = c.imageMetadataStore.Create(*meta); err != nil {
+				return &runtime.PullImageResponse{ImageRef: digest},
+					fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+			}
+		}
+	}
+
+	// Return the image digest
+	return &runtime.PullImageResponse{ImageRef: digest}, err
+}
+
+// imageReferenceResolver attempts to resolve the image reference into a name
+// and manifest via the containerd library call..
+//
+// The argument `ref` should be a scheme-less URI representing the remote.
+// Structurally, it has a host and path. The "host" can be used to directly
+// reference a specific host or be matched against a specific handler.
+//
+// The returned name should be used to identify the referenced entity.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+//
+// If the resolution fails, an error will be returned.
+func (c *criContainerdService) imageReferenceResolver(ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
+	var (
+		ctx  = gocontext.Background()
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler to fetch the object the manifest
+	resolver := docker.NewResolver()
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		var (
+			err     error
+			fetcher remotes.Fetcher
+		)
+
+		resolvedImageName, desc, fetcher, err = resolver.Resolve(ctx, ref)
+		if err != nil {
+			return err
+		}
+
+		eg.Go(func() error {
+			return c.imageStore.Put(ctx, resolvedImageName, desc)
```

Comment:
What does the image store mean? If the image store means the image is on the node, here we only pulled the manifest.

I think for now, we could rely on `pullImage` to pull image, and here we could assume that images are already pulled and the contents are already in store.

We don't need to prepull image metadata for now.